background.js

```diff
@@ -1224,44 +1239,45 @@ const SendLater = {
 
     // Listen for incoming messages, and check if they are in reponse to a scheduled
     // message with a 'cancel-on-reply' header.
-    onNewMailReceivedListener(folder, messagelist) {
+    async onNewMailReceivedListener(folder, messagelist) {
       if (["sent", "trash", "templates", "archives", "junk", "outbox"].includes(folder.type)) {
         SLStatic.debug(`Skipping onNewMailReceived for folder type ${folder.type}`);
         return;
       }
       SLStatic.debug("Received messags in folder", folder, ":", messagelist);
 
       for (let rcvdHdr of messagelist.messages) {
-        messenger.messages.getFull(rcvdHdr.id).then((rcvdMsg) => {
-          SLStatic.debug("Got message", rcvdHdr, rcvdMsg);
-          let inReplyTo = (rcvdMsg.headers["in-reply-to"]||[])[0];
-          if (inReplyTo) {
-            SLTools.forAllDrafts(async (draftHdr) => {
-              if (!SLTools.unscheduledMsgCache.has(draftHdr.id)) {
-                SLStatic.debug(
-                  "Comparing", rcvdHdr, "to", draftHdr,
-                  inReplyTo, "?=", `<${draftHdr.headerMessageId}>`
+        let rcvdMsg = await messenger.messages.getFull(rcvdHdr.id);
```

Comment:
Is there a fear, that the call to getFull() could block the loop? Can we then move the entire body of the loop into an async function, which for each message is properly awaited and throw the returned Promise into an `Promise.all()`?

Reply:
@jph-sendlater would have to comment on why he wrote the code the way he did, and @jobisoft would know better than I whether it's possible for `messenger.messages.getFull` to block, but to the extent that I understand what's going on here, I think the change you propose seems reasonable.
One caveat: if `getFull` could potentially trigger message downloading from the server, then is it a good idea to set off a bunch of those in parallel? Seems like it could cause a spike in network traffic, and also, in my experience, TB sometimes gets a bit unhappy when a bunch of things are all trying to download messages from an IMAP server at the same time. It might therefore be safer to keep the current behavior of processing the messages one by one instead of sticking them all in separate promises and processing them in parallel?

Reply:
It is the opposite. The original code is firing off all operations in the for loop in parallel. My change sequentialized the first `getFull()` request. What I have not yet sequentialized is the call to `SLTools.forAllDrafts()`, that is still scheduled for later async execution without waiting for each to have been finished before the next one is started.

Reply:
I now await the call to `SLTools.forAllDrafts()` below as well.

Reply:
Waiting on feedback from @jph-sendlater for his opinion on the method, before resolving this conversation. I personally feel it is safer to have all run sequentially.